cyral/resource_cyral_access_rules.go

```diff
@@ -0,0 +1,329 @@
+	for _, rule := range arr.AccessRuleSet.AccessRules {
+		m := make(map[string]interface{})
+		m["identity"] = map[string]interface{}{
+			"type": rule.Identity.Type,
+			"name": rule.Identity.Name,
+		}
+		m["validFrom"] = rule.ValidFrom
+		m["validUntil"] = rule.ValidUntil
+		m["config"] = map[string]interface{}{
+			"authorizationPolicyInstanceIDs": rule.Config.AuthorizationPolicyInstanceIDs,
+		}
+
+		rules = append(rules, m)
+	}
+	return d.Set("rule", rules)
+}
+
+// ReadFromSchema is called when *creating* or *updating* a resource.
+// Essentially, it translates the stuff from the .tf file into whatever the
+// API expects. The `AccessRulesResource` will be marshalled verbatim, so
+// make sure that it matches *exactly* what the API needs.
+func (arr *AccessRulesResource) ReadFromSchema(d *schema.ResourceData) error {
+	rules := d.Get("rule").([]interface{})
+	var accessRules []*AccessRule
+
+	for _, rule := range rules {
+		ruleMap := rule.(map[string]interface{})
+
+		accessRule := &AccessRule{}
+
+		identity := ruleMap["identity"].(*schema.Set).List()[0].(map[string]interface{})
+		accessRule.Identity = &AccessRulesIdentity{
+			Type: identity["type"].(string),
+			Name: identity["name"].(string),
+		}
+
+		validFrom := ruleMap["valid_from"].(string)
+		if validFrom != "" {
+			accessRule.ValidFrom = &validFrom
+		}
+
+		validUntil := ruleMap["valid_until"].(string)
+		if validUntil != "" {
+			accessRule.ValidUntil = &validUntil
+		}
+
+		conf, ok := ruleMap["config"]
+		if ok {
+			confList := conf.(*schema.Set).List()
+			if len(confList) > 0 {
+				config := confList[0].(map[string]interface{})
+				policyIDs := config["policy_ids"].([]interface{})
+				ids := make([]string, 0, len(policyIDs))
+				for _, policyID := range policyIDs {
+					ids = append(ids, policyID.(string))
+				}
+				accessRule.Config = &AccessRulesConfig{
+					AuthorizationPolicyInstanceIDs: ids,
+				}
+			}
+		}
+
+		accessRules = append(accessRules, accessRule)
+	}
+
+	arr.AccessRules = accessRules
+	return nil
+}
+
+var ReadRepositoryAccessRulesConfig = ResourceOperationConfig{
+	Name:       "RepositoryAccessRulesRead",
+	HttpMethod: http.MethodGet,
+	CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+		return fmt.Sprintf("https://%s/v1/repos/%s/userAccounts/%s/accessRules",
+			c.ControlPlane,
+			d.Get("repository_id").(string),
+			d.Get("user_account_id").(string),
+		)
+	},
+	NewResponseData: func(_ *schema.ResourceData) ResponseData {
+		return &AccessRulesResponse{}
+	},
+}
+
+func resourceRepositoryAccessRules() *schema.Resource {
+	return &schema.Resource{
+		Description: "Manage access rules",
+		CreateContext: CreateResource(
+			ResourceOperationConfig{
+				Name:       "RepositoryAccessRulesCreate",
+				HttpMethod: http.MethodPost,
+				CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+					repoID := d.Get("repository_id").(string)
+					userAccountID := d.Get("user_account_id").(string)
+					d.SetId(
```

Comment:
The resource ID is usually set after the request succeeds on creating the resource. This can be done in the WriteToSchema of the `AccessRulesResponse`.

Reply:
The problem here is that the creation API does not return an ID, or any of the pieces we need to compute it. The only place where we have the full information is here (that's why Alex suggested I could set the ID here, IIRC).

Reply:
Ah, I keep forgetting what things we have to extract from the API call and which ones we get "for free". I'll fix this, thanks for the suggestion.